### ISSUE TITLE
passing -foreground by default can cause browser hangs

### ISSUE THREAD

**Issue by stuart-little** [mozilla/geckodriver#1888]
## System

* Version: geckodriver 0.29.0 
* Platform: `Linux x86_64`
* Firefox: Mozilla Firefox 88.0.1 
* Selenium: `Python`'s [ selenium 3.141.0 ](https://pypi.org/project/selenium/) package (latest as I file this) 

## Testcase

Simply try this, in a terminal, on a `Linux` machine:

```
firefox -P <PROFILE-NAME> -foreground www.google.com
```
It will cause a window to open, but that window will *not* visit `google.com`. Rather, it will stay frozen on the profile's homepage (for my own it is blank).

Now, this is about Firefox itself, but it can be triggered with `Selenium` (and hence `geckodriver`) as follows:

```
#!/usr/bin/env python

from selenium.webdriver import Firefox, DesiredCapabilities
from selenium.webdriver.firefox.options import Options

options = Options()
options.add_argument("-profile")
options.add_argument("<PATH-TO-EXISTING-PROFILE>")

webdriver = Firefox(desired_capabilities=options.to_capabilities())
webdriver.get("https://google.com")
```
Once more, this opens a window that fails to reach `google.com`.

## Client stacktrace

Because `google.com` is never reached, I have to manually interrupt the process with a `SIGINT`. The trace I get (from `Python`'s `Selenium`) is 
```
^CTraceback (most recent call last):
  File "$HOME/bin/py.py", line 248, in <module>
    webdriver = Firefox(options=opts)
  File "$HOME/.local/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 170, in __init__
    RemoteWebDriver.__init__(
  File "$HOME/.local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 157, in __init__
    self.start_session(capabilities, browser_profile)
  File "$HOME/.local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 252, in start_session
    response = self.execute(Command.NEW_SESSION, parameters)
  File "$HOME/.local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 319, in execute
    response = self.command_executor.execute(driver_command, params)
  File "$HOME/.local/lib/python3.8/site-packages/selenium/webdriver/remote/remote_connection.py", line 374, in execute
    return self._request(command_info[0], url, body=data)
  File "$HOME/.local/lib/python3.8/site-packages/selenium/webdriver/remote/remote_connection.py", line 397, in _request
    resp = self._conn.request(method, url, body=body, headers=headers)
  File "/usr/lib/python3/dist-packages/urllib3/request.py", line 79, in request
    return self.request_encode_body(
  File "/usr/lib/python3/dist-packages/urllib3/request.py", line 171, in request_encode_body
    return self.urlopen(method, url, **extra_kw)
  File "/usr/lib/python3/dist-packages/urllib3/poolmanager.py", line 330, in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.8/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.8/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.8/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
KeyboardInterrupt
```
(also pasted [here](https://termbin.com/fp3p), but that link is bound to expire).

## geckodriver.log

```
1620992605749	geckodriver	INFO	Listening on 127.0.0.1:58315
1620992605752	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "--marionette" "-profile" "$HOME/.mozilla/firefox/ashk27mx.throw" "-foreground" "-no-remote"
Exiting due to channel error.
Exiting due to channel error.
Exiting due to channel error.
```

## Discussion

I originally [filed](https://github.com/SeleniumHQ/selenium/issues/9476#issuecomment-841169264) this with `Selenium`, but was told there that the `-foreground` option is passed by `geckodriver`, not `Selenium`.

The `-foreground` command-line option seems to have been deprecated by `Mozilla`: it is nowhere to be found in `man firefox` or `firefox --help`, and there are precious few references online:

- [This](https://stackoverflow.com/q/47095128) SO discussion [links](https://developer.mozilla.org/en-US/docs/Mozilla/Command_Line_Options#-foreground) to an unmaintained `Mozilla` documentation page.
- On the other hand, [this bug report](https://bugzilla.mozilla.org/show_bug.cgi?id=1415329) on `-foreground` has the following as its closing comment:

> --foreground is an OSX command line argument, we don't have it implemented for other platforms.

One of the earlier comments on there also mentions the option missing from `firefox --help`. 

In conclusion, `-foreground` appears to me to be an abandoned, undocumented option (at least away from `OSX`), and `geckodriver` should not be passing it by default.

**Comment by whimboo** [comment]
Thanks for the report and the late reply. I can give some answers to the behavior you observed:

1. This command line argument is basically only used on Mac and shouldn't have any effect on other platforms. I can see that it prevents navigation on Linux. Would you mind filing a bug for this particular issue? A good place for that should be [Startup and Profile System component on Bugzilla](https://bugzilla.mozilla.org/enter_bug.cgi?product=Toolkit&component=Startup%20and%20Profile%20System).
2. Details to why it's not listed are getting discussed at https://bugzilla.mozilla.org/show_bug.cgi?id=1415329.
3. We can stop setting this argument in geckodriver when the current platform is not MacOS. I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1720502 for this particular issue.

Let me know if you have other questions, and please inform us when you filed the bug in Bugzilla. Thanks!

**Comment by whimboo** [comment]
The code in geckodriver has been updated to only set this argument on MacOS. You can try out a [nightly build of geckodriver](https://treeherder.mozilla.org/jobs?repo=mozilla-central&revision=225f9de7bf4e7818dd01e06bbcad67400a89a2fb&searchStr=gd-s) to verify that it now works.

Again feel free to file a separate bug for Firefox as mentioned above.

Closing issue now given that the work on geckodriver is done.
